azurekvp/src/kvp.rs

```diff
@@ -0,0 +1,503 @@
+use std::fs::OpenOptions;
+use std::io::{self, Error, ErrorKind, Write};
+use std::os::fd::AsRawFd;
+use std::os::unix::fs::MetadataExt;
+use std::path::Path;
+
+const HV_KVP_EXCHANGE_MAX_KEY_SIZE: usize = 512;
+const HV_KVP_EXCHANGE_MAX_VALUE_SIZE: usize = 2048;
+const HV_KVP_AZURE_MAX_VALUE_SIZE: usize = 1024;
+const EVENT_PREFIX: &str = concat!("azure-init-", env!("CARGO_PKG_VERSION"));
+
+/// A custom visitor that captures the value of the `msg` field from a tracing event.
+/// It implements the `tracing::field::Visit` trait and records the value into
+/// a provided mutable string reference.
+///
+/// This visitor is primarily used in the `on_event` method of the `EmitKVPLayer`
+/// to extract event messages and log them as key-value pairs.
+pub struct StringVisitor<'a> {
+    string: &'a mut String,
+}
+
+impl<'a> Visit for StringVisitor<'a> {
+    /// Records the debug representation of the event's value and stores it in the provided string.
+    ///
+    /// # Arguments
+    /// * `_field` - A reference to the event's field metadata.
+    /// * `value` - The debug value associated with the field.
+    fn record_debug(
+        &mut self,
+        _field: &tracing::field::Field,
+        value: &dyn std_fmt::Debug,
+    ) {
+        write!(self.string, "{:?}", value).unwrap();
+    }
+}
+/// A custom tracing layer that emits span data as key-value pairs (KVP)
+/// to a file for consumption by the Hyper-V daemon. This struct captures
+/// spans and events from the tracing system and writes them to a
+/// specified file in KVP format.
+pub struct EmitKVPLayer {
+    file_path: std::path::PathBuf,
+}
+
+impl EmitKVPLayer {
+    pub fn new(file_path: std::path::PathBuf) -> Self {
+        Self { file_path }
+    }
+}
+
+impl<S> Layer<S> for EmitKVPLayer
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup>,
+{
+    /// Handles event occurrences within a span by extracting the `msg` field,
+    /// recording the event, and writing the event as key-value pairs (KVP) to a log file.
+    ///
+    /// This method uses the `StringVisitor` to capture the message of the event and
+    /// links the event to the span by writing both span and event data into the same file.
+    ///
+    /// If an `ERROR` level event is encountered, a failure flag is inserted into the span's extensions.
+    ///
+    /// # Example
+    /// ```rust
+    /// use tracing::{event, Level};
+    /// event!(Level::INFO, msg = "Event message");
+    /// ```
+    fn on_event(&self, event: &tracing::Event<'_>, ctx: TracingContext<'_, S>) {
+        let mut event_message = String::new();
+
+        let mut visitor = StringVisitor {
+            string: &mut event_message,
+        };
+
+        event.record(&mut visitor);
+
+        if let Some(span) = ctx.lookup_current() {
+            let mut extensions = span.extensions_mut();
+
+            if event.metadata().level() == &tracing::Level::ERROR {
+                extensions.insert("failure".to_string());
```

Comment:
Similar to the `Instant` extension, we don't want to use `String` as the type, we'll want a new private type. Additionally, the new type could be an `enum` instead of wrapping a string. Or, if "failure" is the only state (and the absence of the extension signals not-failure) you could define an empty struct and use that: `struct Failure;`